### PR TITLE
tracee: init at 0.7.0

### PR DESCRIPTION
**Pull request by 06kellyjac**
###### Description of changes

Package tracee at ~`0.6.5`~ `0.7.0` for nixpkgs

~I tried building from source but had some issues :upside_down_face:~
~I got it to play ball by using `clang` wrapped until I can't, then using `clang-unwrapped`, then back to `clang` wrapped~
I use use normal wrapped clang, then a slightly modified wrapped clang called clang-bpf, then normal wrapped clang

to test `tracee-ebpf`:

```
$ sudo ./result/bin/tracee-ebpf \
	-o option:parse-arguments \
	--trace comm=bash \
	--trace follow \
	--trace event!='sched*'
```

to test that + `tracee-rules`:

```
$ sudo ./result/bin/tracee-ebpf \
	-o format:json \
	-o option:parse-arguments \
	--trace comm=bash \
	--trace follow \
	--trace event!='sched*' | \
	./result/bin/tracee-rules \
	--input-tracee file:stdin \
	--input-tracee format:json \
	--rules-dir ./result/share/tracee/rules

# you should expect

Loaded 14 signature(s): [TRC-1 TRC-13 TRC-2 TRC-14 TRC-3 TRC-11 TRC-9 TRC-4 TRC-5 TRC-12 TRC-8 TRC-6 TRC-10 TRC-7]

# then run within a bash shell: sudo strace ls
# you should expect

*** Detection ***
Time: 2022-03-29T20:57:42Z
Signature ID: TRC-2
Signature: Anti-Debugging
Data: map[]
Command: strace
Hostname: YOURHOSTNAME
```

You should also have `/sys/kernel/btf/vmlinux` (`ls /sys/kernel/btf/vmlinux` to check)

<https://aquasecurity.github.io/tracee/v0.7.0/install/prerequisites/>

**Note: I've put rules in `usr/share/tracee/rules` but tracee-rules can automatically use rules directly next to the binary. We could move them but I guess they're best opt in?**

```
result
├── bin
│   ├── tracee-ebpf
│   └── tracee-rules
└── share
    └── tracee
        ├── rules
        │   ├── anti_debugging_ptraceme.rego
        │   ├── builtin.so
        │   ├── cgroup_release_agent_modification.rego
        │   ├── code_injection.rego
        │   ├── disk_mount.rego
        │   ├── dropped_executable.rego
        │   ├── dynamic_code_loading.rego
        │   ├── fileless_execution.rego
        │   ├── helpers.rego
        │   ├── illegitimate_shell.rego
        │   ├── k8s_service_account_token.rego
        │   ├── kernel_module_loading.rego
        │   ├── kubernetes_certificate_theft_attempt.rego
        │   └── ld_preload.rego
        └── templates
            ├── csv.tmpl
            ├── falcosidekick.tmpl
            ├── rawjson.tmpl
            ├── simple.tmpl
            ├── sprig.tmpl
            └── xml.tmpl

```

###### Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [X] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [X] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [X] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by 06kellyjac** [comment]
Result of `nixpkgs-review pr 163477` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>tracee</li>
  </ul>
</details>

File: pkgs/tools/security/tracee/default.nix
Comment (papojari) [inline pkgs/tools/security/tracee/default.nix:40]: That site doesn't exist.
```suggestion
    homepage = "https://aquasecurity.github.io/tracee/latest";
```
This one does.

File: pkgs/tools/security/tracee/default.nix
Comment (papojari) [inline pkgs/tools/security/tracee/default.nix:52]: The readme only says *Linux*. Nothing about x86_64.
```suggestion
    platforms = platforms.linux;
```

File: pkgs/tools/security/tracee/default.nix
Comment (papojari) [inline pkgs/tools/security/tracee/default.nix:3]: Wouldn't it be better to use `buildGoModule`?

**Review by auroraanna — CHANGES_REQUESTED** [review]

File: pkgs/tools/security/tracee/default.nix
Comment (06kellyjac) [reply pkgs/tools/security/tracee/default.nix:52]: Since im fetching from a prebuilt release thats only built for x86_64 linux it has to be pinned

File: pkgs/tools/security/tracee/default.nix
Comment (06kellyjac) [reply pkgs/tools/security/tracee/default.nix:3]: Ive tried building from source and it's been a nightmare because its all designed to be built within their dev containers.
This derivation just patches the prebuilt release so stdenvNoCC.mkDerivation is good.


For a source build I'll almost certainly be using buildGoModule

File: pkgs/tools/security/tracee/default.nix
Comment (papojari) [reply pkgs/tools/security/tracee/default.nix:52]: Another reason to use `buildGoModule` so we have support for multiple architectures ^^

File: pkgs/tools/security/tracee/default.nix
Comment (06kellyjac) [reply pkgs/tools/security/tracee/default.nix:40]: I was trying that link out but yeah it didnt work. I didnt mean to commit that one

**Review by auroraanna — APPROVED** [review]

**Comment by papojari** [comment]
##### Reviewed points

- [X] package path fits guidelines
- [ ] package name fits guidelines
- [ ] package version fits guidelines
- [ ] package build on x86_64-linux
- [ ] executables tested on x86_64-linux
- [ ] `meta.description` is set and fits guidelines
- [ ] `meta.license` fits upstream license
- [ ] `meta.platforms` is set
- [ ] `meta.maintainers` is set
- [ ] phases are respected

**Comment by papojari** [comment]
[ ] made sure ofBorg build succeeded for all applicable platforms

**Comment by 06kellyjac** [comment]
I've made some progress on the source build. It works but its very messy so I'm looking to clean it up.
It might take a while so I dont mind if this gets merged in the meanwhile.

**Comment by 06kellyjac** [comment]
I've swapped to draft and added the source build on top for now.
Once I'm happy with it I'll squash down to 1 commit

**Comment by 06kellyjac** [comment]
It didn't build on my aarch64 box. I'm not going to delve any deeper into this while there are comments like this:

https://github.com/aquasecurity/tracee/blob/bc05bc0f9caf3e7d32ea8021c28fb6256ee3a799/pkg/ebpf/c/tracee.bpf.c#L3180

image

File: pkgs/tools/security/tracee/default.nix
Comment (SuperSandro2000) [inline pkgs/tools/security/tracee/default.nix:86]:
```suggestion
```

Comment (06kellyjac) [reply pkgs/tools/security/tracee/default.nix:86]: I'll be swapping it to just x86_64 linux as I had no luck with aarch64

**Comment by 06kellyjac** [comment]
waiting for the next release soon to clean this up and undraft

**Comment by 06kellyjac** [comment]
RC is out but waiting for 0.7.0 full release

File: pkgs/tools/security/tracee/default.nix
Comment (SuperSandro2000) [inline pkgs/tools/security/tracee/default.nix:61]: Missing makeFlags which includes -j$NIX_BUILD_CORES

File: pkgs/tools/security/tracee/default.nix
Comment (SuperSandro2000) [inline pkgs/tools/security/tracee/default.nix:59]: We can just drop https://github.com/aquasecurity/tracee/blob/main/Makefile#L313=

File: pkgs/tools/security/tracee/default.nix
Comment (SuperSandro2000) [inline pkgs/tools/security/tracee/default.nix:35]: Can't we use our libbpf?

File: pkgs/tools/security/tracee/default.nix
Comment (06kellyjac) [reply pkgs/tools/security/tracee/default.nix:61]: Ive seen some rust builds within stdenv.mkDeriviation before so might be able to do it that way around

Comment (SuperSandro2000) [reply pkgs/tools/security/tracee/default.nix:61]: I don't know how easy it is to carry over the go vendor stuff but you should be able to just add ``$makeFlags``.

File: pkgs/tools/security/tracee/default.nix
Comment (06kellyjac) [reply pkgs/tools/security/tracee/default.nix:35]: I think I got it to work but I was thinking it might be best if the libbpf version was updated in tandem with tracee.

I guess we could use our libbpf until it breaks.
I might write a nixos test to make sure it stays working.
A normal test probably wont play ball since itll need sudo and can probably trace outside the sandbox

Comment (06kellyjac) [reply pkgs/tools/security/tracee/default.nix:35]: Ok I tried using `libbpf.a` directly from nixpkgs but I couldn't really get it working

I also tried just copying the `libbpf.src` into `3rdparty/libbpf` but it wasn't very happy about that either

I think part of the issue is the whole `$(OUT_DIR)/whatever:` stuff but to be fair I'm not an avid make user

File: pkgs/tools/security/tracee/default.nix
Comment (06kellyjac) [reply pkgs/tools/security/tracee/default.nix]: I've handled it with `"CMD_GIT=echo"`

I want to avoid changing the makefile too much to reduce maintenance burden

File: pkgs/tools/security/tracee/default.nix
Comment (SuperSandro2000) [reply pkgs/tools/security/tracee/default.nix:77]: I wanted to get rid of bash arrays and so on but well...

Comment (06kellyjac) [reply pkgs/tools/security/tracee/default.nix:77]: I can probably trim it down a bit.
I guess i could completely drop the array stuff but it might be handy for overriding rather than having to replace the whole buildphase

File: pkgs/tools/security/tracee/default.nix
Comment (rafaeldtinoco) [reply pkgs/tools/security/tracee/default.nix:35]: Please don't use your own libbpf. We test features per libbpf release (and per libbpfgo release). It is better if you use our builtin version just like we do in our packaging.

Comment (06kellyjac) [reply pkgs/tools/security/tracee/default.nix:35]: Cool, thanks for the review

Comment (SuperSandro2000) [reply pkgs/tools/security/tracee/default.nix:35]: You are [pinning libbpf version 0.7.0](https://github.com/aquasecurity/tracee/tree/v0.7.0/3rdparty) which is the exact version we currently have. There are no patches and vendoring in this case has only disadvantages for us: We rely on upstream to update it and patches from our libbpf are not applied. It might break when other libraries are updated and wouldn't be fixed. Also vendored libraries increase compile times because we build the lib more often than necessary.

Comment (rafaeldtinoco) [reply pkgs/tools/security/tracee/default.nix:35]: @SuperSandro2000 I know about the cons of having a static lib redistributed and that is not something I would request from any other project usually (I'm an Ubuntu core dev). Unfortunately, libbpf isn't stable until 1.0 (they try to be, but there might be breaking changes until they're fully stable) AND we have faced issues from 0.5 to 0.7, for example. So, let's say next release (0.7.1, or 0.8) of tracee still depends on libbpf0.7 + libbpfgo0.7+xxx and you may need to update libbpf to 0.8 to follow upstream, because of others tools (like bpftool), then we would start having a problem. If your users come to our github to open an issue we would request them to drop your package and use ours (instead of recommending them to use yours like we could do). At the end, it is your decision obviously, it was just a kind request. Thanks!

Comment (06kellyjac) [reply pkgs/tools/security/tracee/default.nix:35]: Ok, these are the two approaches Im happy with:

1. Keep it as is at least until libbpf hits 1.0
2. Use the nixpkgs libbpf but I'll put some checks to make sure it matches the pinned version tracee wants. If the main nixpkgs libbpf goes ahead Ill pin a separate one at the right version

My preference is for option 1 since its less overhead and a minor compromise for a higher guarantee of tracee working.

Option 2 is quite a lot of work and at that point Id probably go with an option 3 which would be - just use the tracee release bin like the first draft of this PR, where we'd go from building 90% ideal solution to a less ideal one

Comment (rafaeldtinoco) [reply pkgs/tools/security/tracee/default.nix:35]: I like the approach of having static libbpf for tracee until libbpf is 1.0. I think that, from point on, if we find an issue in tracee because of libbpf then it will be on us either way. Thanks for considering it.

Comment (SuperSandro2000) [reply pkgs/tools/security/tracee/default.nix:35]: yeah, fine with me.

File: pkgs/tools/security/tracee/default.nix
Comment (06kellyjac) [reply pkgs/tools/security/tracee/default.nix]: I've trimmed this down to just the makeflags part + enableParallelBuilding

I could drop `enableParallelBuilding` and just hardcode the `-j` and `-l` flags but I think this is fine

**Comment by 06kellyjac** [comment]
I think most things here have been addressed now

@SuperSandro2000 I'd love to get your thoughts on the nixos test (https://github.com/06kellyjac/nixpkgs/blob/tracee-integration/pkgs/tools/security/tracee/test.nix). Is this something I should hold-off of adding to passthru and just run manually until RFC0119 is finalized/libbpf hits 1.0 or just add it now?

**Comment by 06kellyjac** [comment]
Ok, I've rebased and I think all comments are resolved now. I added the nixosTest but it's not in passthru, just to be manually ran for now. I'm hoping to get this in before 22.05 ideally

**Comment by SuperSandro2000** [comment]
> Is this something I should hold-off of adding to passthru and just run manually until RFC0119 is finalized/libbpf hits 1.0 or just add it now?

IDK.

**Pull request merged**
